dlt/common/cli/runner/pipeline_script.py

```diff
@@ -0,0 +1,145 @@
+import os
+import importlib
+import typing as t
+import sys
+
+from contextlib import contextmanager
+from types import ModuleType
+
+import dlt
+
+from dlt.cli import echo as fmt
+from dlt.cli.utils import parse_init_script
+from dlt.common.cli.runner.errors import RunnerError
+from dlt.common.cli.runner.types import PipelineMembers, RunnerParams
+from dlt.common.pipeline import LoadInfo
+from dlt.sources import DltResource, DltSource
+
+
+COMMAND_NAME: t.Final[str] = "run"
+
+
+class PipelineScript:
+    """Handles pipeline source code and prepares it to run
+
+    Things it does
+
+    1. Reads the source code,
+    2. Stubs all pipeline.run calls,
+    3. Creates a temporary module in the location of pipeline file,
+    4. Imports rewritten module,
+    5. Prepares module name,
+    6. Exposes ready to use module instance.
+    """
+
+    def __init__(self, params: RunnerParams) -> None:
+        self.params = params
+        """This means that user didn't specify pipeline name or resource and source name
+
+        And we need to check if there is 1 pipeline and 1 resource or source to run right
+        away if there are multiple resources and sources then we need to provide a CLI prompt
+        """
+        self.workdir = os.path.dirname(os.path.abspath(params.script_path))
+        """Directory in which pipeline script lives"""
+
+        self.has_pipeline_auto_runs: bool = False
+
+        # Now we need to patch and store pipeline code
+        visitor = parse_init_script(
+            COMMAND_NAME,
+            self.script_contents,
+            self.module_name,
+        )
+        self.source_code = visitor.source
+
+    def load_module(self, script_path: str) -> ModuleType:
+        """Loads pipeline module from a given location"""
+        with self.expect_no_pipeline_runs():
+            spec = importlib.util.spec_from_file_location(self.module_name, script_path)
+            module = importlib.util.module_from_spec(spec)
+            sys.modules[self.module_name] = module
+            spec.loader.exec_module(module)
+            if self.has_pipeline_auto_runs:
+                raise RunnerError(
+                    "Please move all pipeline.run calls inside __main__ or remove them"
+                )
+
+            return module
+
+    @contextmanager
+    def expect_no_pipeline_runs(self):  # type: ignore[no-untyped-def]
+        """Monkey patch pipeline.run during module loading
+        Restore it once importing is done
+        """
+        old_run = dlt.Pipeline.run
+
+        def noop(*args, **kwargs) -> LoadInfo:  # type: ignore
+            self.has_pipeline_auto_runs = True
+
+        dlt.Pipeline.run = noop  # type: ignore
+
+        yield
+
+        dlt.Pipeline.run = old_run  # type: ignore
```

Comment:
@rudolfix I followed this way to `noop` patch pipeline.run to detect if pipeline has been run immediately in the script

Reply:
And on line [#62](https://github.com/dlt-hub/dlt/pull/1045/files#diff-a6d60aa16741e8f3f79da1202df5678420e8b70a67d2733bb8dfe25d710f88b2R62) I exit with error message